### PR TITLE
Fix for dynamic + static args on log

### LINKED ISSUE #1206: Logs are Malformed if Functions are Called in It
### Version Information

* vyper Version: `0.1.0b7`
* OS: `Ubuntu 18.04`
* Python Version (python --version): `3.6.7`
* Environment (output of `pip freeze`): 
```
pkg-resources==0.0.0
pycryptodome==3.7.2
vyper==0.1.0b7
```

### What's your issue about?

Calling a function when emitting a log leads to a malformed log. However, a normal 
log is emitted when an object is placed in the log instead of a function.

Please include information like:

* full output of the error you received
This test is in `go` however this error is relevant as when unpacking and decoding logs emitted, I do get the below error.

```
--- FAIL: TestProcessChainStartLog (1.63s)
    log.go:31: log: Unable to unpack ChainStart log data unable to unpack logs: abi: cannot marshal in to go slice: offset 3321006415 would go over slice boundary (len=128)
```

* the code that caused the failure (see [this link](https://help.github.com/articles/basic-writing-and-formatting-syntax/) for help with formatting code)

The below code causes an error:

```py 

TestLog: event({testData: bytes[8] })

@private
@constant
def to_bytes(value: uint256) -> bytes[8]:
    return slice(concat("", convert(value, bytes32)), start=24, len=8)

@public
def test_func(value: uint256):

  # log will be malformed
   TestLog(self.to_bytes(value))
   
   loggedValue: bytes[8] = self.to_bytes(value)
   
   # log will be normal
   TestLog(loggedValue)

```

### How can it be fixed?

Fill this in if you know how to fix it.

#### Cute Animal Picture
image

### PR DESCRIPTION
**Pull request by jacqueswww**
### - What I did
Fixes #1206.

### CODE REVIEW

File: tests/parser/features/test_logging.py [hunk 1]
```diff
@@ -868,3 +868,43 @@ def set_list():
...
+
+    logs = get_logs(tx_hash, c, 'TestLog')
+
+    assert logs[0].args == logs[1].args
+
+    log = logs[0].args
+    assert w3.toInt(log.testData1) == 123
+    assert w3.toInt(log.testData2[:32]) == 123
+    assert log.testData2[-7:] == b'testing'
+    assert log.testData2[32:64] == b'\x00\x00\x00\x00\x00\x00\x00{testing'
```
Comment (nisdas) [inline tests/parser/features/test_logging.py]: Wouldnt this be out of bounds with the array having a size of 60 ?

Comment (jacqueswww) [reply tests/parser/features/test_logging.py]: ah right, python still handles > max, so the assert is still correct ;)

Comment (jacqueswww) [reply tests/parser/features/test_logging.py]:
```
>>> a ="aaaa"
>>> a[1:100]
'aaa'
```

**Comment by jacqueswww** [comment]
@nisdas Could you test your geth test implementation on this branch? If all looks in order I can merge.

**Comment by nisdas** [comment]
@jacqueswww Hey, so I still have failing tests 

This is what passes and fails, at least now dynamic args kinda work , 
like in the second case the log is normal and I am able to decode and use,

```py
 TestLog: event({testData1: bytes32,testData2: bytes[60], testData3: bytes[8] })

@public
@constant
def to_bytes(value: uint256) -> bytes[8]:
    return slice(concat("", convert(value, bytes32)), start=24, len=8)

@private
@ constant
def to_bytes32(value: uint256) -> bytes32:
    return convert(value, bytes32)

@public
def test_func(value: uint256):

data2: bytes[2064] = concat(self.to_bytes32(value),self.to_bytes(value),"testing")

  # log will be malformed
   log.TestLog(self.to_bytes32(value),data2,self.to_bytes(value))
   
   loggedValue: bytes[8] = self.to_bytes(value)

   # log will be normal
   log.TestLog(self.to_bytes32(value)  ,data2,loggedValue)
```

however in the first case, having a dynamic arg for the third data type leads
to the second data type being corrupted. On my end my tests indicate that 
that the second data type that was logged isnt what was expected.
 When we use a static arg instead the log is normal and all tests pass

**Comment by nisdas** [comment]
For reference this is the actual contract I am testing on now 
```py 
## compiled with v0.1.0-beta.7 ##

DEPOSIT_CONTRACT_TREE_DEPTH: constant(uint256) = 32
TWO_TO_POWER_OF_TREE_DEPTH: constant(uint256) = 4294967296  # 2**32
SECONDS_PER_DAY: constant(uint256) = 86400

Deposit: event({previous_deposit_root: bytes32, data: bytes[2064], merkle_tree_index: bytes[8]})
ChainStart: event({deposit_root: bytes32, time: bytes[8]})

MIN_DEPOSIT_AMOUNT: uint256
MAX_DEPOSIT_AMOUNT: uint256
CHAIN_START_FULL_DEPOSIT_THRESHOLD: uint256
deposit_tree: map(uint256, bytes32)
deposit_count: uint256
full_deposit_count: uint256

@public
def __init__(depositThreshold: uint256,minDeposit: uint256,maxDeposit: uint256):
    self.CHAIN_START_FULL_DEPOSIT_THRESHOLD = depositThreshold
    self.MIN_DEPOSIT_AMOUNT = minDeposit
    self.MAX_DEPOSIT_AMOUNT = maxDeposit


@private
@constant
def to_bytes(value: uint256) -> bytes[8]:
    return slice(concat("", convert(value, bytes32)), start=24, len=8)

@public
@constant
def get_deposit_root() -> bytes32:
    return self.deposit_tree[1]

@payable
@public
def deposit(deposit_input: bytes[2048]):
    deposit_amount: uint256 = msg.value / as_wei_value(1, "gwei")

    assert deposit_amount >= self.MIN_DEPOSIT_AMOUNT
    assert deposit_amount <= self.MAX_DEPOSIT_AMOUNT

    deposit_timestamp: uint256 = as_unitless_number(block.timestamp)
    deposit_data: bytes[2064] = concat(self.to_bytes(deposit_amount), self.to_bytes(deposit_timestamp), deposit_input)
    index: uint256 = self.deposit_count + TWO_TO_POWER_OF_TREE_DEPTH
    merkle_tree_index: bytes[8] = self.to_bytes(index)

    log.Deposit(self.get_deposit_root(), deposit_data, merkle_tree_index)

    # Add deposit to merkle tree
    self.deposit_tree[index] = sha3(deposit_data)
    for i in range(DEPOSIT_CONTRACT_TREE_DEPTH):
        index /= 2
        self.deposit_tree[index] = sha3(concat(self.deposit_tree[index * 2], self.deposit_tree[index * 2 + 1]))

    self.deposit_count += 1
    if deposit_amount == self.MAX_DEPOSIT_AMOUNT:
        self.full_deposit_count += 1
        if self.full_deposit_count == self.CHAIN_START_FULL_DEPOSIT_THRESHOLD:
            timestamp_day_boundary: uint256 = deposit_timestamp - deposit_timestamp % SECONDS_PER_DAY + SECONDS_PER_DAY
            log.ChainStart(self.get_deposit_root(), self.to_bytes(timestamp_day_boundary))

@public
@constant
def get_branch(leaf: uint256) -> bytes32[DEPOSIT_CONTRACT_TREE_DEPTH]:
    branch: bytes32[32] # size is DEPOSIT_CONTRACT_TREE_DEPTH
    index: uint256 = leaf + TWO_TO_POWER_OF_TREE_DEPTH
    for i in range(DEPOSIT_CONTRACT_TREE_DEPTH):
        branch[i] = self.deposit_tree[bitwise_xor(index, 1)]
        index /= 2
    return branch
```
This contract passes all my tests , however if I have my deposit log written like this 
```py
   log.Deposit(self.get_deposit_root(), deposit_data, self.to_bytes(index))
``` 
then my tests fail

**Comment by nisdas** [comment]
@jacqueswww hey all the tests on my end pass :)

**Comment by jacqueswww** [comment]
@nisdas Great.

So looking at the deposit contract I would highly not recommend creating `to_bytes` function.  Considering how large the local scope is because of the `data2` array all the context will have to be popped onto the stack before going to private call (possible limit + huge gas costs). So I would either just inline it or make it a public function.

**Comment by jacqueswww** [comment]
Is the bigger problem perhaps that you'd want slice to work like so?
`a: bytes[8] = slice(value ,start=24, len=8)`
I think we have an issue for this somewhere - I can pick up to get that behaviour.

**Comment by JustinDrake** [comment]
> So I would either just inline it or make it a public function.

Making it a public function works for me :)

**Comment by nisdas** [comment]
@jacqueswww Using a public function would work best for that purpose, since we are looking not to duplicate code across the contract. Yeah if it was bytes32 it would be simpler for us and we wouldnt need to give a dynamic argument in the log. Do you see any other areas in the contract that could be made more efficient ? I am very new to vyper, so I am still familiarizing myself with it

**Comment by jacqueswww** [comment]
That was the only thin I picked up at the moment, I have created https://github.com/ethereum/vyper/issues/1219 to add support for bytes slicing a bytes32 type.

**Pull request merged**